osatreesitter/docgen.py

```diff
@@ -0,0 +1,412 @@
+            'structure' dictionary and generating a formatted string.
+
+        count_tokens(prompt, model)
+            Counts the number of tokens in a given prompt using a specified model.
+
+        generate_class_documentation(class_details, model)
+            Generates documentation for a class using OpenAI GPT.
+
+        generate_method_documentation()
+            Generates documentation for a single method using OpenAI GPT.
+
+        extract_pure_docstring(gpt_response)
+            Extracts only the docstring from the GPT-4 response while keeping triple quotes.
+
+        insert_docstring_in_code(source_code, method_details, generated_docstring)
+            Inserts a generated docstring into the specified location in the source code.
+
+        insert_cls_docstring_in_code(source_code, class_details, generated_docstring)
+            Inserts a generated class docstring into the class definition and returns the updated source code.
+
+        process_python_file(parsed_structure, file_path)
+            Processes a Python file by generating and inserting missing docstrings and updates the source file
+            with the new docstrings.
+
+        generate_documentation_openai(file_structure, model)
+            Generates the documentation for a given file structure using OpenAI's API by traversing the given
+            file structure and for each class or standalone function, generating its documentation.
+    """
+
+    def __init__(self):
+        """
+        Instantiates the object of the class.
+
+        This method is a constructor that initializes the object by setting the 'api_key' attribute to the value of the 'OPENAI_API_KEY' environment variable.
+
+        Args:
+            self (object): A reference to the instance of the class.
+
+        Returns:
+            None
+        """
+        self.api_key = os.getenv("OPENAI_API_KEY")
+
+    @staticmethod
+    def format_structure_openai(structure: dict):
+        """
+        Formats the structure of Python files in a readable string format.
+
+        This method iterates over the given dictionary 'structure' and generates a formatted string where it describes
+        each file, its classes and functions along with their details such as line number, arguments, return type,
+        source code and docstrings if available.
+
+        Args:
+            structure (dict): A dictionary containing details of the Python files structure. The dictionary should
+            have filenames as keys and values as lists of dictionaries. Each dictionary in the list represents a
+            class or function and should contain keys like 'type', 'name', 'start_line', 'docstring', 'methods'
+            (for classes), 'details' (for functions) etc. Each 'methods' or 'details' is also a dictionary that
+            includes detailed information about the method or function.
+
+        Returns:
+            str: A formatted string representing the structure of the Python files.
```

Comment:
Типы данных достаточно указывать в определении функции, в докстрингах не надо 
```suggestion
            structure: A dictionary containing details of the Python files structure. The dictionary should
            have filenames as keys and values as lists of dictionaries. Each dictionary in the list represents a
            class or function and should contain keys like 'type', 'name', 'start_line', 'docstring', 'methods'
            (for classes), 'details' (for functions) etc. Each 'methods' or 'details' is also a dictionary that
            includes detailed information about the method or function.

        Returns:
            A formatted string representing the structure of the Python files.
```